In [1]:
cd ../..

D:\PyCharmProject\nna_24-25_pham


In [2]:
import numpy as np
import torch
import matplotlib.pyplot as plt
from torch.utils.data import DataLoader, TensorDataset
from data.generators import *
from utils.trainer import train_model
from utils.evaluator import evaluate_model
from utils.visualizer import *
from utils.maths import *
from model.simple_model import ReLU_Network
from typing import Callable

In [3]:
FUNCTION = lambda x: x**2
A = 0
B = 1

In [4]:
error = compute_approximation_error(f = FUNCTION, interval=(A,B))
error["error"]

0.125

### Explanation for the error $\Delta(S^*)$
The explanation for the error $\Delta(S^*)$ being $0.125$ is provided in the `experiments/conditions.ipynb` notebook. Readers are encouraged to refer to the notebook for a detailed explanation.

In [5]:
assert error['c'] == (B+A)/2
print("C is the midpoint of the interval [A, B], as expected.")

assert error['d'] == (error['c'] + A)/2
print("D is the midpoint between C and A, as expected.")

C is the midpoint of the interval [A, B], as expected.
D is the midpoint between C and A, as expected.



We observe that the optimal approximation occurs at $C$, which is the midpoint of the interval $[A, B]$.
This is because the second derivative of the function $f(x) = x^2$ is constant and equal to $2$.

Mathematically, this can be expressed as:
$$
f''(x) = 	constant
$$

For a quadratic function, $f(x)$ can be written in terms of its second derivative as:
$$
f(x) = [0.5 \cdot f''(x)] \cdot x^2 + k \cdot x + e,
$$
where $k$ and $e$ are constants.

The derivative at the midpoint $c$ can be calculated as:
$$
f'(c) = \frac{f(b) - f(a)}{b - a}.
$$

Substituting the values of $f(b)$ and $f(a)$ into the equation:
$$
f'(c) = \frac{\left( \frac{f'' \cdot b^2}{2} + k \cdot b \right) - \left( \frac{f'' \cdot a^2}{2} + k \cdot a \right)}{b - a}.
$$

Simplifying further:
$$
f'(c) = \frac{f'' \cdot (b-a) \cdot (b+a)}{2(b-a)} + k.
$$

This reduces to:
$$
f'(c) = \frac{f'' \cdot (b+a)}{2} + k = f'' \cdot c + k.
$$

This implies that $c = \frac{a+b}{2}$ is the optimal approximation for the function $f(x)$ when $f''(x)$ is constant.

Similarly, it can be shown that $d = \frac{c+a}{2}$ is the midpoint between $c$ and $a$, providing further approximation points.

## Proof of Theorem 2: Bounds on Approximation Error

We observe that the optimal approximation error occurs when the function has constant second-order derivatives.
Let $\bar{f}(x)$ represent the upper-bound approximation of the function $f(x)$, where the second derivative
of $\bar{f}(x)$ is constant and equal to $\max_{x \in [a,b]} f''(x)$.
Similarly, let $\underline{f}(x)$ represent the lower-bound approximation, where the second derivative is 
constant and equal to $\min_{x \in [a,b]} f''(x)$.

### Upper Bound
The upper bound of $\Delta(f_n^*)$ is derived as follows:
$$
\Delta(f_n^*) \leq \Delta(\bar{f}_n^*) \leq \frac{(b-a)^2}{16 \cdot n^2} \cdot \max_{x \in [a,b]} f''(x).
$$

This is because $c_i$, the point of optimal approximation, lies at the midpoint of each segment $[a_i, b_i]$. Mathematically:
$$
c_i - a_i = \frac{b_i - a_i}{2}, \quad d_i - a_i = \frac{b_i - a_i}{4}.
$$

Substituting these values:
$$
\Delta(S_i^*) \leq \frac{b_i - a_i}{4} \cdot \bar{f}'' \cdot \left[ \frac{b_i - a_i}{2} - \frac{b_i - a_i}{4} \right] = \frac{(b_i - a_i)^2}{16} \cdot \max_{x \in [a,b]} f''(x).
$$

Summing across $n$ equal segments ($b_i - a_i = \frac{b-a}{n}$):
$$
\Delta(\bar{f}_n^*) \leq \frac{(b-a)^2}{16 \cdot n^2} \cdot \max_{x \in [a,b]} f''(x).
$$

### Lower Bound
Similarly, the lower bound is:
$$
\Delta(f_n^*) \geq \Delta(\underline{f}_n^*) \geq \frac{(b-a)^2}{16 \cdot n^2} \cdot \min_{x \in [a,b]} f''(x).
$$

### Final Result
Combining both bounds, we conclude:
$$
\frac{(b-a)^2}{16 \cdot n^2} \cdot \min_{x \in [a,b]} f''(x) \leq \Delta(f_n^*) \leq \frac{(b-a)^2}{16 \cdot n^2} \cdot \max_{x \in [a,b]} f''(x).
$$

This completes the proof of Theorem 2.

### Convergence Rate
The convergence rate of the error $\Delta(f_n^*)$ is $\mathcal{O}(n^{-2})$. This implies that the error decreases quadratically with the number of segments $n$.

Since $b-a$ is irrelevant to the convergence rate while being the interval length, the error is independent of the interval length. Meanwhile, minimal and maximal second derivatives are crucial for the error bounds, also not affecting the convergence rate.

Therefore, the error $\Delta(f_n^*)$ is independent of the interval length and converges quadratically with the number of segments $n$. In other words, the Convergence Rate is $\mathcal{O}(n^{-2})$.